File: 10866/10866.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    10866.py                                           :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/03/02 16:12:43 by jihoolee          #+#    #+#              #
#    Updated: 2024/03/02 16:18:49 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

import sys
from collections import deque


input = sys.stdin.readline


class Deque(deque):
    def __init__(self):
        super().__init__()

    def popleft(self) -> int:
        if len(self) == 0:
            return -1
        return super().popleft()

    def pop(self) -> int:
        if len(self) == 0:
            return -1
        return super().pop()

    def left(self) -> int:
        if len(self) == 0:
            return -1
        return self[0]

    def right(self) -> int:
        if len(self) == 0:
            return -1
        return self[-1]


def main() -> None:
    d: Deque = Deque()

    for _ in range(int(input())):
        cmd: list[str] = input().split()

        if cmd[0] == 'push_front':
            d.appendleft(int(cmd[1]))
        elif cmd[0] == 'push_back':
            d.append(int(cmd[1]))
        elif cmd[0] == 'pop_front':
            print(d.popleft())
        elif cmd[0] == 'pop_back':
            print(d.pop())
        elif cmd[0] == 'size':
            print(len(d))
        elif cmd[0] == 'empty':
            print(int(len(d) == 0))
        elif cmd[0] == 'front':
            print(d.left())
        elif cmd[0] == 'back':
            print(d.right())


if __name__ == '__main__':
    main()

File: 10866/10866.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   10866.cpp                                          :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/03/02 16:07:59 by jihoolee          #+#    #+#             */
/*   Updated: 2024/03/02 16:11:36 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <string>
#include <deque>

int pop_front_deque(std::deque<int>* p_d) {
  if (p_d->empty())
    return -1;

  int result = p_d->front();

  p_d->pop_front();
  return result;
}

int pop_back_deque(std::deque<int>* p_d) {
  if (p_d->empty())
    return -1;

    int result = p_d->back();

    p_d->pop_back();
    return result;
}

int get_front_deque(std::deque<int>* p_d) {
  if (p_d->empty())
    return -1;
  return p_d->front();
}

int get_back_deque(std::deque<int>* p_d) {
  if (p_d->empty())
    return -1;
  return p_d->back();
}

void do_cmd(std::deque<int>* p_d, const std::string& cmd) {
  int num;

  if (cmd == "push_front") {
    std::cin >> num;
    p_d->push_front(num);
  } else if (cmd == "push_back") {
    std::cin >> num;
    p_d->push_back(num);
  } else if (cmd == "pop_front") {
    std::cout << pop_front_deque(p_d) << std::endl;
  } else if (cmd == "pop_back") {
    std::cout << pop_back_deque(p_d) << std::endl;
  } else if (cmd == "size") {
    std::cout << p_d->size() << std::endl;
  } else if (cmd == "empty") {
    std::cout << int(p_d->empty()) << std::endl;
  } else if (cmd == "front") {
    std::cout << get_front_deque(p_d) << std::endl;
  } else if (cmd == "back") {
    std::cout << get_back_deque(p_d) << std::endl;
  }
}

int main(void) {
  int             cmd_cnt;
  std::string     cmd;
  std::deque<int> d;

  std::cin >> cmd_cnt;
  for (int i = 0; i < cmd_cnt; ++i) {
    std::cin >> cmd;
    do_cmd(&d, cmd);
  }
  return 0;
}
